core/src/main/scala/chisel3/internal/Builder.scala

```diff
@@ -210,15 +210,24 @@ private[chisel3] trait HasId extends InstanceId {
   // Uses a namespace to convert suggestion into a true name
   // Will not do any naming if the reference already assigned.
   // (e.g. tried to suggest a name to part of a Record)
-  private[chisel3] def forceName(prefix: Option[String], default: => String, namespace: Namespace): Unit =
+  private[chisel3] def forceName(
+    prefix:     Option[String],
+    default:    => String,
+    namespace:  Namespace,
+    errorIfDup: Boolean = false,
+    refBuilder: String => Arg = (x: String) => (Ref(x))
+  ): Unit =
     if (_ref.isEmpty) {
-      val candidate_name = _computeName(prefix, Some(default)).get
+      val candidate_name = _computeName(prefix, Some(default).filterNot(_ => errorIfDup)).get
```

Comment:
I know i suggested this but that was just for development, this logic is not very clear as to what it's trying to do. Notably, the result of `_computeName` can only ever been `None` if the thing has no name and there's no `default`. If we were to expose that `default` is an `Option`, we could change this to a `.getOrElse` (with no `filterNot`) and fold the error from `checkPorts` into here (and make sure to pass `None` as the default for ports).